feat(a380x): us units option

MikioDK opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
Fixes #9206 

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->

- Added conversion for FUEL&LOAD page entries.
- Displaying values on Fuel/Apu/Engine/Cruise SD pages in the correct unit
- Displaying correct unit on OANS page
- Displaying correct unit for T.O Shift on PERF TO page
- Displaying correct unit on F-Plan for Departure/approach runway
- Fixed some UI components to be more efficient and a lot easier to use.
- Fixed some type issues with NXDataStore.
- Converted CONFIG_USING_METRIC to the new style setting.

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make your best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->

## References
<!-- You should be making changes based on some kind of a reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

![image](https://github.com/user-attachments/assets/b3416eba-7080-4615-bdbe-43cd89593e52)

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub): Mikio

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->

- Switch weight unit on Flypad Aircraft setting, enter values as LBS on Fuel & Load page. Cycle through SD pages of APU/Engine/Fuel/Cruise to check for correct unit of values.
- Select Departure Runway, Select Arrival Runway on F-PLAN and check it shows correct unit.
- Use OANS and BTV, and select runway and exit and check it shows correct unit.

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo**, **flybywire-aircraft-a380-842 (4K)** or **flybywire-aircraft-a380-842 (8K)** download link at the bottom of the page

Benjozork commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnArr.tsx:
```diff
@@ -7,6 +7,7 @@ import { ApproachType, LandingSystemUtils } from '@fmgc/index';
 import { getApproachName } from '../../../shared/utils';
 
 import './MfdFmsFpln.scss';
+import { NXUnits } from '@flybywiresim/fbw-sdk';
```
**style:** move this to line 2

Benjozork commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnArr.tsx:
```diff
@@ -454,9 +455,9 @@ export class MfdFmsFplnArr extends FmsPage<MfdFmsFplnArrProps> {
                   }}
                 >
-                  {this.rwyLength}
+                  {NXUnits.mToUser(this.rwyLength)}
```
**question:** I'm not sure if I'm missing something, but I don't think this works. `NXUnits.mToUse` takes in a `number`, and you are passing in a `Subject<number>`.

I think we need to approach this differently. Even if this worked, right now, it would not react to the EFB setting changing, potentially requiring restarting your flight.

I would suggest looking into the `NumberUnit` / `NumberUnitSubject` types from `msfs-sdk`, and properly hooking that to react with the kg/lbs setting from `NXDataStore`.

MikioDK replied:
- That is correct, have made a correction to it.

in regards to the approach, it does work to do the conversion directly in the onNewData() method inside the component.

It does not however work in real-time with the EFB setting changing, i.e you need to reload the entire page, by going out of it and into it again, **however this is the case with all the other NXUnit conversions as well.** so if it has to happen in real time, then it needs to be used for all the areas.

tracernz replied:
I don't think that would be very nice. A better approach might be something wrapping or inheriting NumberUnitSubject and that can hide all the details of listening for the setting, and performing the conversions.

MikioDK replied:
If anything, it would be subscribing to NXDataStore out in MFDPageDirectory and the passing that value down, so everything refreshes when it changes. Then we never need to ask if its metric on NXUnit, and we can then make etc. `NumberUnit(value, KG).asUnit(passedDownUnit)` and call it a day.

Benjozork replied:
The UI does not react to things that are not `Subscribable` values. I'm not sure what you mean. So there is value in `NumberUnitSubject`.

It also even has its own `asUnit` that takes a subscribable of a unit - allowing you to refresh the UI when the unit changes.

MikioDK committed:
Fixed issue with RWY length conversion

MikioDK committed:
Fixed issue with FOB unit conversion

MikioDK committed:
Typos and simplified refactor

MikioDK commented:
The change now means that there is no reference to NXUnits from the MFD or ND, everything is now a MSFSSDK Unit.

I have not found a nice way to prevent having to rebuild the list of runway dropdown options just to change to label on them, suppose thats the price to pay.

I have not really found a suitable place to put the Unit fomart string helpers, to write out "KG", "T", "LBS" etc. right now they just live in each component. I suppose it could live in some static class somewhere.

I am confident that it works on OANS as well, but haven't found a way to check it in dev. At least its made exactly the same as the DataFormatEntry for InputField.

tracernz commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnArr.tsx:
```diff
@@ -394,6 +423,21 @@ export class MfdFmsFplnArr extends FmsPage<MfdFmsFplnArrProps> {
       }, true),
     );
+
+    NXDataStore.getAndSubscribe('CONFIG_USING_METRIC_UNIT', (key, value) => {
+      value === '1' ? this.unit.set(UnitType.METER) : this.unit.set(UnitType.FOOT);
+
+      if (!this.props.fmcService.master || !this.loadedFlightPlan) {
+        return;
+      }
+
+      const isAltn = this.props.fmcService.master.revisedWaypointIsAltn.get() ?? false;
+      const flightPlan = isAltn ? this.loadedFlightPlan.alternateFlightPlan : this.loadedFlightPlan;
+
+      if (flightPlan.destinationAirport) {
+        this.GenerateRunwayOptions(flightPlan, isAltn);
+      }
+    });
```
I'd do all the other work in a sub on lengthUnit for cleanliness. METER needs to be the default, both for invalid garbage in the setting, hence inverting the condition, and if it's not set at all (the 3rd param for getAndSubscribe).
```suggestion
    NXDataStore.getAndSubscribe('CONFIG_USING_METRIC_UNIT', (key, value) => {
      this.lengthUnit.set(value === '0' ? UnitType.FOOT : UnitType.METER);
    }, '1');
      
    this.lengthUnit.sub(() => {
      if (!this.props.fmcService.master || !this.loadedFlightPlan) {
        return;
      }

      const isAltn = this.props.fmcService.master.revisedWaypointIsAltn.get() ?? false;
      const flightPlan = isAltn ? this.loadedFlightPlan.alternateFlightPlan : this.loadedFlightPlan;

      if (flightPlan.destinationAirport) {
        this.GenerateRunwayOptions(flightPlan, isAltn);
      }
    }, true);
```

tracernz commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnDep.tsx:
```diff
@@ -190,6 +222,17 @@ export class MfdFmsFplnDep extends FmsPage<MfdFmsFplnDepProps> {
+
+      if (flightPlan?.destinationAirport) {
+        this.GenerateRunwayOptions(flightPlan, isAltn);
+      }
```
As above for the arr page.

tracernz commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnArr.tsx:
```diff
@@ -79,6 +91,10 @@ export class MfdFmsFplnArr extends FmsPage<MfdFmsFplnArrProps> {
 
   private apprButtonScrollTo = Subject.create<number>(0);
 
+  private unit = Subject.create<Unit<UnitFamily.Distance>>(
+    NXDataStore.get('CONFIG_USING_METRIC_UNIT') === '1' ? UnitType.METER : UnitType.FOOT,
+  );
```
We immediately get the value after render below, so to save having to duplicate the logic and default value, just set metre initially, and also a slightly better name.
```suggestion
  private lengthUnit = Subject.create<Unit<UnitFamily.Distance>>(UnitType.METER);
```

tracernz commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnDep.tsx:
```diff
@@ -40,43 +52,20 @@ export class MfdFmsFplnDep extends FmsPage<MfdFmsFplnDepProps> {
 
+  private unit = Subject.create<Unit<UnitFamily.Distance>>(
+    NXDataStore.get('CONFIG_USING_METRIC_UNIT') === '1' ? UnitType.METER : UnitType.FOOT,
+  );
```
As above for arr page.

tracernz commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/MfdFmsPerf.tsx:
```diff
@@ -415,6 +425,10 @@ export class MfdFmsPerf extends FmsPage<MfdFmsPerfProps> {
 
   private missedEngineOutAccelAltIsPilotEntered = Subject.create<boolean>(false);
 
+  private unit = Subject.create<Unit<UnitFamily.Distance>>(
+    NXDataStore.get('CONFIG_USING_METRIC_UNIT') === '1' ? UnitType.METER : UnitType.FOOT,
+  );
```
```suggestion
  private lengthUnit = Subject.create<Unit<UnitFamily.Distance>>(UnitType.METER);
```

MikioDK replied:
I did this inititially, and it means that it will always be Meter on first render, even if you have set LBS on Flypad, as the subscribe is onAfterRender() and no value updates to set the LengthUnit differently

MikioDK replied:
Its unclear to me if this will still be a potential thing or not

Benjozork replied:
> as the subscribe is onAfterRender() and thhere are no value updates, to set the LengthUnit differently before you click on the field.

`.sub` has a second parameter which lets you have it update immediately

tracernz commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/MfdFmsPerf.tsx:
```diff
@@ -963,6 +977,10 @@ export class MfdFmsPerf extends FmsPage<MfdFmsPerfProps> {
         }),
     );
+
+    NXDataStore.getAndSubscribe('CONFIG_USING_METRIC_UNIT', (key, value) => {
+      value === '1' ? this.unit.set(UnitType.METER) : this.unit.set(UnitType.FOOT);
+    });
```
```suggestion
    NXDataStore.getAndSubscribe('CONFIG_USING_METRIC_UNIT', (key, value) => {
      this.lengthUnit.set(value === '0' ? UnitType.FOOT : UnitType.METER);
    }, '1');
```

MikioDK replied:
I did this initially, and the side-effect is that if you have lbs set on the flypad before navigating to the page, then it will always be in Meters. As the subscribe comes "onAfterRender".

tracernz commented on fbw-a380x/src/systems/instruments/src/ND/OansControlPanel.tsx:
```diff
@@ -170,6 +173,10 @@ export class OansControlPanel extends DisplayComponent<OansProps> {
 
   public interactionMode = Subject.create<InteractionMode>(InteractionMode.Touchscreen);
 
+  private unit = Subject.create<Unit<UnitFamily.Distance>>(
+    NXDataStore.get('CONFIG_USING_METRIC_UNIT') === '1' ? UnitType.METER : UnitType.FOOT,
+  );
```
```suggestion
  private lengthUnit = Subject.create<Unit<UnitFamily.Distance>>(UnitType.METER);
```

tracernz commented on fbw-a380x/src/systems/instruments/src/ND/OansControlPanel.tsx:
```diff
@@ -218,6 +225,10 @@ export class OansControlPanel extends DisplayComponent<OansProps> {
 
     NXDataStore.getAndSubscribe('NAVIGRAPH_ACCESS_TOKEN', () => this.loadOansDb());
 
+    NXDataStore.getAndSubscribe('CONFIG_USING_METRIC_UNIT', (key, value) => {
+      value === '1' ? this.unit.set(UnitType.METER) : this.unit.set(UnitType.FOOT);
+    });
```
```suggestion
    NXDataStore.getAndSubscribe('CONFIG_USING_METRIC_UNIT', (key, value) => {
      this.lengthUnit.set(value === '0' ? UnitType.FOOT : UnitType.METER);
    }, '1');
```

tracernz commented on fbw-common/src/systems/instruments/src/ND/ND.tsx:
```diff
@@ -151,6 +155,10 @@ export class NDComponent<T extends number> extends DisplayComponent<NDProps<T>>
 
   private showOansRunwayInfo = Subject.create<boolean>(false);
 
+  private unit = Subject.create<Unit<UnitFamily.Distance>>(
+    NXDataStore.get('CONFIG_USING_METRIC_UNIT') === '1' ? UnitType.METER : UnitType.FOOT,
+  );
```
```suggestion
  private lengthUnit = Subject.create<Unit<UnitFamily.Distance>>(UnitType.METER);
```

tracernz commented on fbw-common/src/systems/instruments/src/ND/ND.tsx:
```diff
@@ -228,6 +236,10 @@ export class NDComponent<T extends number> extends DisplayComponent<NDProps<T>>
       .whenChanged()
       .handle((show) => this.showOans.set(show));
+
+    NXDataStore.getAndSubscribe('CONFIG_USING_METRIC_UNIT', (key, value) => {
+      value === '1' ? this.unit.set(UnitType.METER) : this.unit.set(UnitType.FOOT);
+    });
```
```suggestion
    NXDataStore.getAndSubscribe('CONFIG_USING_METRIC_UNIT', (key, value) => {
      this.lengthUnit.set(value === '0' ? UnitType.FOOT : UnitType.METER);
    }, '1');
```

MikioDK commented:
ping review?

Benjozork commented:
> ping review?

There are remaining review comments that are not closed or responded to

jbud commented:
Is there any assistance needed with this? I'd love to help with any remaining items

MikioDK committed:
style changes

MikioDK committed:
Initialize measurements as metric

MikioDK committed:
cleanup

MikioDK committed:
cleanup

MikioDK committed:
Added unit conversion to support lbs/ft instead of kg/m

MikioDK committed:
Added changelog entry

MikioDK committed:
Moved way from NXUnits onto MSFSSDK unit to be subscribeable

MikioDK committed:
rebase on master

MikioDK committed:
Fixed merge issue

MikioDK committed:
Fixed merge issue

tracernz committed:
Merge remote-tracking branch 'upstream/master' into pr/9432

tracernz committed:
fix: input field units

tracernz committed:
fix: fuel&load page

tracernz committed:
feat: buttons optionally take subscribable as label

tracernz committed:
chore: mark todo items

tracernz committed:
feat: convert metric setting to modern style

tracernz committed:
fix: work around type unsoundness issue

tracernz committed:
chore: actually, this is worth the cost for how often it might happen

tracernz committed:
fix(a380x): button component

Get rid of the Subscribable<VNode> label as that should never be needed
and it adds a bunch of unnecessary complexity. Add Subscribable<string> label.

tracernz committed:
fix: rwy formatting

tracernz committed:
feat: support us units on airport data page

tracernz committed:
feat: fpln us unit

tracernz committed:
feat: appr us unit

tracernz committed:
feat: hold us unit

tracernz committed:
docs: ammend changelog

tracernz committed:
Merge remote-tracking branch 'upstream/master' into a380-lbs-fms-inputs

tracernz commented:
This is RTR.

BravoMike99 commented on fbw-a380x/src/systems/instruments/src/MFD/pages/common/DataEntryFormats.tsx:
```diff
@@ -382,28 +391,41 @@ export class LengthFormat extends SubscriptionCollector implements DataEntryForm
 
   private maxValue = Number.POSITIVE_INFINITY;
 
+  public unitFamily?: Subscribable<Unit<UnitFamily.Distance>> = Subject.create(UnitType.METER);
```
Couldn't we just set it in the constructor like in the weight case?

tracernz replied:
The weight one actually doesn't need the logic to handle no unitFamily since it has a default arg...
Note that this code actually does get compiled into the constructor.

BravoMike99 commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/DATA/MfdFmsDataAirport.tsx:
```diff
@@ -67,13 +69,9 @@ export class MfdFmsDataAirport extends FmsPage<MfdFmsDataAirportProps> {
       FSComponent.render(
         <Button
-          label={runway.label}
+          label={this.useMetric.map((v) => MfdFmsDataAirport.formatRunwayLabel(runway, v, false))}
```
Don't we need to destroy this one?

BravoMike99 commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFpln.tsx:
```diff
@@ -53,6 +61,16 @@ export interface DerivedFplnLegData {
 }
 
 export class MfdFmsFpln extends FmsPage<MfdFmsFplnProps> {
+  private readonly weightUnit = NXDataStore.getSetting('CONFIG_USING_METRIC_UNIT').map((v) =>
```
Similar regarding this one.

BravoMike99 commented on fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnHold.tsx:
```diff
@@ -10,11 +18,21 @@ import { InputField } from 'instruments/src/MsfsAvionicsCommon/UiWidgets/InputFi
 interface MfdFmsFplnHoldProps extends AbstractMfdPageProps {}
 
 export class MfdFmsFplnHold extends FmsPage<MfdFmsFplnHoldProps> {
+  private readonly weightUnit = NXDataStore.getSetting('CONFIG_USING_METRIC_UNIT').map((v) =>
```
another one?

BravoMike99 commented on fbw-a380x/src/systems/instruments/src/ND/OansControlPanel.tsx:
```diff
@@ -234,6 +235,10 @@ export class OansControlPanel extends DisplayComponent<OansProps> {
 
   public interactionMode = Subject.create<InteractionMode>(InteractionMode.Touchscreen);
 
+  private readonly lengthUnit = NXDataStore.getSetting('CONFIG_USING_METRIC_UNIT').map((v) =>
```
+1?

BravoMike99 reviewed (commented):
Some smaller comments, I can  handle the merge conflicts and apply the changes if needed after confirmation of these.

tracernz committed:
Merge remote-tracking branch 'upstream/master' into mrc-trk-brg

tracernz committed:
a380x(fms): make unitFamily non-optional

tracernz committed:
a380x(fms/oanc): handle lifecycles manually

tracernz committed:
a380x(fms): fix placeholders

tracernz committed:
a380x(fms): fix tow/lw units

tracernz committed:
a380x(fms): convert perf dest efob

tracernz committed:
a380x(fms): fix fuel page efob and zfw placeholder

tracernz committed:
a380x(fms): fix runway length format and convert the info value

tracernz committed:
a380x(fms): fixes for master merge

tracernz committed:
Merge remote-tracking branch 'upstream/master' into a380-lbs-fms-inputs

BravoMike99 commented:
<!-- !! DO NOT REMOVE THIS LINE !! - QUALITY ASSURANCE REPORT -->
<!-- Refer to the https://docs.flybywiresim.com/dev-corner/qa-process/#reporting-on-tests for details about this template -->

<!-- Please use one or the other depending on your Discord Role -->
Dev team QA report

Discord Username  : bruno_pt99
Object of testing : #9432 
Aircraft          : A380X 
Simulator         : 2020
Tier of Testing   : 2
Date              : 15/04/2026

Testing Process: 
Cycled through the EFB setting and checked that the MFD pages display in the correct unit (perf, fuel & load, F-PLN, DATA/AIRPORT, HOLD, F-PLN DEP/ARR, F-PLN) ✅ .
Checked that the SD pages also display the correct unit ✅ .
Did a quick pattern to setup BTV with the manual fallback (no navi subscription) and noticed the units were not converted and the buttons are not rendered properly. ❌ 

Testing Results: 
Not Passed

Negatives:       
Units are not converted in the BTV fallback window.
Airport selection buttons of the OANS are not rendered correctly.
SD ENG page unit text is not updated if the unit is changed while the page is open.

Media:
<img width="643" height="728" alt="SD" src="https://github.com/user-attachments/assets/ed94c631-73c8-48fb-b074-bdff4df253fe" />
<img width="735" height="264" alt="oans" src="https://github.com/user-attachments/assets/64c2666f-3be9-4d6b-813c-d99f749279d0" />

tracernz commented:
~~The OANC is intentionally out of scope as we are not sure it even has a US units option, and there was some suspicion that it doesn't. I updated the PR text to reflect that.~~
-e- I am lazy and it's over half done, so I'll do the rest and we can back out the change if we find it's not a thing.

tracernz committed:
a380x(sd): fix units not updating

tracernz committed:
a380x(oans): fix fallback length unit and clean code

tracernz committed:
a380x(oans): fix buttons

tracernz committed:
a380x(oans): improve fallback formatting

tracernz committed:
Merge remote-tracking branch 'upstream/master' into a380-lbs-fms-inputs

BravoMike99 commented:
<!-- !! DO NOT REMOVE THIS LINE !! - QUALITY ASSURANCE REPORT -->
<!-- Refer to the https://docs.flybywiresim.com/dev-corner/qa-process/#reporting-on-tests for details about this template -->

<!-- Please use one or the other depending on your Discord Role -->
Dev team QA report

Discord Username  : bruno_pt99
Object of testing : #9432 
Aircraft          : A380X <!-- Choose only one -->
Simulator         : 2020 <!-- Choose only one -->
Tier of Testing   : 2
Date              : 16/04/2026

Testing Process: 
Performed a quick pattern around LPBJ.
Checked that MFD pages react accordingly to the EFB units.
Checked that SD pages react accordingly to the EFB units.
Verified issues raised on the last test no longer occur.
ND BTV fallback displays accordingly to the EFB units (did not check OANS BTV as don't have navi ultimate subscription).

Testing Results: 
Passed

Negatives:
CRZ SD says "LB" instead of "LBS" and "LB/H" instead of "LBS/H".
Weight unit is missing on the FF all engines on the fuel page.
<img width="575" height="733" alt="Cruise" src="https://github.com/user-attachments/assets/057ae18c-16fd-4fa2-bf12-a6d66bad9331" />

Media:
<img width="565" height="737" alt="SD fuel" src="https://github.com/user-attachments/assets/5cf6235d-2846-49a7-89ce-af51930eaa21" />
<img width="567" height="721" alt="SD eng" src="https://github.com/user-attachments/assets/8cb5a72c-a7f5-4f63-8a86-cbc155f17a31" />
<img width="1599" height="995" alt="Unit2" src="https://github.com/user-attachments/assets/72ca30bb-66ba-4330-8460-8b26b9b9a9b6" />
<img width="1617" height="997" alt="Units1" src="https://github.com/user-attachments/assets/137bc9a8-f254-4aa0-8d73-fb0385ebf281" />
<img width="735" height="288" alt="BTV" src="https://github.com/user-attachments/assets/d2f34b97-22b6-47a3-989f-a0cf3c6b872d" />

BravoMike99 commented on fbw-a380x/src/systems/instruments/src/SD/Pages/FuelPage.tsx:
```diff
@@ -1046,9 +1045,8 @@ export const FuelPage = () => {
-      {/* TODO unit switching? */}
       <text x={68} y={644} className="Cyan T2">
-        KG/MIN
+        {NXUnits.userWeightUnit}/MIN
```
missing a ()?

tracernz committed:
a380x(sd): LB->LBS

tracernz committed:
a380x(sd): fuel unit
